### PR TITLE
Refactor persistent worker key computation

### PR DESCRIPTION
**Pull request by ulfjack**
Extract the computation of the persistent worker key into a helper
class. This makes it easier to test and reuse (e.g., for remote
persistent workers).

Change-Id: Ia9aa073423281b67c16b4a351b13a812d9849e1c

### CODE REVIEW

**Comment by ulfjack** [comment]
@larsrc-google, here's the promised PR.

**Comment by larsrc-google** [comment]
Thanks! A good idea. Quick comment since I'm about to leave for the evening: Could you remove the things from WorkerSpawnRunner that now live in WorkerParser? That'll make it easier to check that this doesn't change functionality.

**Comment by ulfjack** [comment]
Done. Sorry I missed that.

File: src/main/java/com/google/devtools/build/lib/worker/WorkerParser.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        .filter(entry -> entry.getKey().equals(spawn.getMnemonic()))
+        .forEach(entry -> mnemonicFlags.add(entry.getValue()));
+
+    return workerArgs
+        .add("--persistent_worker")
+        .addAll(MoreObjects.firstNonNull(mnemonicFlags.build(), ImmutableList.of()))
+        .build();
+  }
+
+  private static UserExecException createUserExecException(
```
Comment (larsrc-google) [inline src/main/java/com/google/devtools/build/lib/worker/WorkerParser.java:118]: You're only using this once, no need for a method.

File: src/main/java/com/google/devtools/build/lib/worker/WorkerParser.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.SortedMap;
+import java.util.regex.Pattern;
+
+class WorkerParser {
+  public static final String ERROR_MESSAGE_PREFIX =
+      "Worker strategy cannot execute this %s action, ";
+  public static final String REASON_NO_FLAGFILE =
```
Comment (larsrc-google) [inline src/main/java/com/google/devtools/build/lib/worker/WorkerParser.java:27]: Remove this from WorkerSpawnRunner

File: src/main/java/com/google/devtools/build/lib/worker/WorkerParser.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+      "because the command-line arguments do not contain at least one @flagfile or --flagfile=";
+
+  /** Pattern for @flagfile.txt and --flagfile=flagfile.txt */
+  private static final Pattern FLAG_FILE_PATTERN = Pattern.compile("(?:@|--?flagfile=)(.+)");
```
Comment (larsrc-google) [inline src/main/java/com/google/devtools/build/lib/worker/WorkerParser.java:31]: Remove this from WorkerSpawnRunner, too.

File: src/main/java/com/google/devtools/build/lib/worker/WorkerParser.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+import com.google.devtools.build.lib.vfs.Path;
+import com.google.devtools.build.lib.vfs.PathFragment;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.SortedMap;
+import java.util.regex.Pattern;
+
+class WorkerParser {
```
Comment (larsrc-google) [inline src/main/java/com/google/devtools/build/lib/worker/WorkerParser.java:24]: Please add docs.

File: src/main/java/com/google/devtools/build/lib/worker/WorkerParser.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+      LocalEnvProvider localEnvProvider,
+      BinTools binTools) {
+    this.execRoot = execRoot;
+    this.multiplex = multiplex;
+    this.workerOptions = workerOptions;
+    this.localEnvProvider = localEnvProvider;
+    this.binTools = binTools;
+  }
+
+  public WorkerConfig compute(Spawn spawn, SpawnRunner.SpawnExecutionContext context)
```
Comment (larsrc-google) [inline src/main/java/com/google/devtools/build/lib/worker/WorkerParser.java:52]: Please add docs.

File: src/main/java/com/google/devtools/build/lib/worker/WorkerParser.java (hunk 1 above)
Comment (ulfjack) [reply src/main/java/com/google/devtools/build/lib/worker/WorkerParser.java:118]: Done.

File: src/main/java/com/google/devtools/build/lib/worker/WorkerParser.java (hunk 4 above)
Comment (ulfjack) [reply src/main/java/com/google/devtools/build/lib/worker/WorkerParser.java:28]: Done.

File: src/main/java/com/google/devtools/build/lib/worker/WorkerParser.java (hunk 2 above)
Comment (ulfjack) [reply src/main/java/com/google/devtools/build/lib/worker/WorkerParser.java:31]: Done.

File: src/main/java/com/google/devtools/build/lib/worker/WorkerParser.java (hunk 3 above)
Comment (ulfjack) [reply src/main/java/com/google/devtools/build/lib/worker/WorkerParser.java:35]: Done.

File: src/main/java/com/google/devtools/build/lib/worker/WorkerParser.java (hunk 5 above)
Comment (ulfjack) [reply src/main/java/com/google/devtools/build/lib/worker/WorkerParser.java]: Done.

**Review by larsrc-google — APPROVED** [review]
LGTM. I'm moving the protocolformat check inside the parser.

**Pull request closed without merging**